### PR TITLE
Refactoring TTIR\TTNN ops cpp files

### PR DESCRIPTION
**Pull request by sdjordjevicTT**
Currently, all our op verifiers and other op extensions(extraClassDeclarations and op interfaces) are implemented in one big cpp file per dialect:
https://github.com/tenstorrent/tt-mlir/blob/main/lib/Dialect/TTIR/IR/TTIROps.cpp
https://github.com/tenstorrent/tt-mlir/blob/main/lib/Dialect/TTNN/IR/TTNNOps.cpp

This change introduces the structure of operations into these files, improving code readability. For example, all op-specific methods(verifier, folder, etc...) are grouped inside the large file. As we add more MLIR features(like op interfaces, external op declarations, op folders, etc.), this change and organization will simplify future code contributions.

Closes https://github.com/tenstorrent/tt-mlir/issues/896
Closes https://github.com/tenstorrent/tt-mlir/issues/888

**Edit**: We agreed to keep the single file, but we will use this PR to give those single files a structure.

### CODE REVIEW

**Review by nsmithtt — COMMENTED** [review]
Hey Stefan, next time can we run refactors like this by each other before doing it?

I think I'd personally prefer to keep all the ops in one place, I don't really see a huge benefit to breaking them out.  This adds a lot of boilerplate for adding a new op and breaks up their verify functions which I prefer to have all in the same place.  It's also non-conventional, all other dialects I've checked, Arith, Math, Tosa, Linalg, torch-mlir, stablehlo all have the ops in one file.

If we do go with this:
- Can we use LLVM file naming convention, i.e. Titlecase?
- Can we at least group the ops into categories? I think it's too fine grain to have every op have its own include and cpp file, for e.g. adding an eltwise op should just be adding a few lines instead of adding new files and updating cmake etc.

**Comment by sdjordjevicTT** [comment]
> Hey Stefan, next time can we run refactors like this by each other before doing it?
> 
> I think I'd personally prefer to keep all the ops in one place, I don't really see a huge benefit to breaking them out. This adds a lot of boilerplate for adding a new op and breaks up their verify functions which I prefer to have all in the same place. It's also non-conventional, all other dialects I've checked, Arith, Math, Tosa, Linalg, torch-mlir, stablehlo all have the ops in one file.
> 
> I'd like to get others thoughts on this and understand what this buys us. IMO large files aren't necessarily bad as long as they encapsulate related code, i.e. the IR ops.
> 
> If we do go with this:
> 
> * Can we use LLVM file naming convention, i.e. Titlecase?
> * Can we at least group the ops into categories? I think it's too fine grain to have every op have its own include and cpp file, for e.g. adding an eltwise op should just be adding a few lines instead of adding new files and updating cmake etc.

Yes, it's a bit much, but the current state of the single file is not ideal either. Every new operation function is added in a random spot, and as we continue adding more ops and features to the dialect, it only gets messier. That's why I'd like to refactor it sooner rather than later.

I have reviewed other dialects, and while most of them also use a single file, I have come across a few exceptions that take a different approach (like ONNX-MLIR and, to some extent, the SPIRV dialect).

I agree, as majority of dialects keeps them in the single file, we should keep them as well. I can use this PR to give single files a bit of the structure. From what I have seen, many dialects (e.g., LLVM IR, Torch, StableHLO) organize their files by operation, while TOSA groups them by functionality (verifiers, shape inferences, etc.). I lean towards the first approach, where files are divided by ops, as it seems cleaner to me, but I am open to other suggestions if you have any ideas.

**Comment by rpavlovicTT** [comment]
> This is one of those topics where it's a personal preference thing. So I'll share mine :)
> 
> I don't like either of the extremes - all ops in one file is hard to keep organized, so ops that are semantically similar/relevant tend to be randomly distributed within the file; each op in its own file feels like a hassle to deal with for reasons @nsmithtt points out.
> 
> If I get a vote, it's for organizing ops by "functionality" (tensormanipulation, binary, unary, tensorcreation, matmuls ...)
> 
> `</bikeshed>`

First of all I agree with Nick that splitting everything into files doesn't help much. And it is nice if we keep up with standards other dialects put before us. 

Second, if we kept code in a single file I am more in favor of grouping code by op type. Just like Stefan suggested. 

Third, even within a single file we can have nice organization. For example, we can add more namespaces which can play role of a file hierarchy. 
```
namespace data_movement {
...
}
namespace eltwise {
...
}
...
```

**Comment by nsmithtt** [comment]
> I agree, as majority of dialects keeps them in the single file, we should keep them as well. I can use this PR to give single files a bit of the structure.
>
> Second, if we kept code in a single file I am more in favor of grouping code by op type. Just like Stefan suggested.
>
> Third, even within a single file we can have nice organization. For example, we can add more namespaces which can play role of a file hierarchy.

+1 for give the file more structure!  I vote for this approach, by op type makes sense.

**Comment by sdjordjevicTT** [comment]
Great, I will update the PR description and send an updated version to reflect our agreement. Thanks, everyone for the input!

**Comment by sdjordjevicTT** [comment]
I updated the refactor, please take a look at the reorganized files and give me some feedback. :)

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 1]
```diff
@@ -2,258 +2,71 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-#include <cstdint>
-#include <string>
+#include "ttmlir/Dialect/TTIR/IR/TTIROps.h"
```
Comment (svuckovicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp:5]: It stills seems a little random. For example, I would've expected to see conv2d and maxpool ops near each other. Can we group the ops by their functionality?

E.g.:
- memory/layout ops (tolayout, tomemcfg, todevice, fromdevice, ...)
- tensor creation ops (empty, full, ones, zeros, ...)
- eltwise unary ops
- eltwise binary ops
- matmul ops
- reduction ops
- conv ops (conv2d, maxpool, avgpool, ...)
- TM ops (transpose, concat, reshape, slice, ...)

Comment (sdjordjevicTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:5]: > It stills seems a little random. For example, I would've expected to see conv2d and maxpool ops near each other. Can we group the ops by their functionality?
> 
> E.g.:
> 
> * memory/layout ops (tolayout, tomemcfg, todevice, fromdevice, ...)
> * tensor creation ops (empty, full, ones, zeros, ...)
> * eltwise unary ops
> * eltwise binary ops
> * matmul ops
> * reduction ops
> * conv ops (conv2d, maxpool, avgpool, ...)
> * TM ops (transpose, concat, reshape, slice, ...)

That was the idea :) Currently, ops are divided in TTIR:
- Conv
- DataMovement(TMs)
- Layout
- Matmul
- MemoryManagement(alloc)
- Normalization(Softmax)
- Pooling
- Generic
- Kernel

I will merge pooling under the Conv category, that makes sense. Groups are in alphabetic order except Generic\Kernel, they seem to be related, hence I put them at the end. Maybe it makes sense to extract Generic and Kernel ops implementation into a separate file...

Comment (svuckovicTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:5]: Ah, I guess I first went to check for conv&maxpool, thanks for letting me know!

**Review by svuckovicTT — APPROVED** [review]

**Review by nsmithtt — APPROVED** [review]

**Pull request merged**